u8timeseries/models/tcn_model.py

```diff
@@ -0,0 +1,254 @@
+        x += residual
+
+        return x
+
+
+class _TCNModule(nn.Module):
+    def __init__(self,
+                 input_size: int,
+                 input_length: int,
+                 kernel_size: int,
+                 num_filters: int,
+                 num_layers: Optional[int],
+                 dilation_base: int,
+                 weight_norm: bool,
+                 output_length: int,
+                 dropout: float):
+
+        """ PyTorch module implementing a dilated TCN module used in `TCNModel`.
+
+
+        Parameters
+        ----------
+        input_size
+            The dimensionality of the input time series.
+        input_length
+            The length of the input time series.
+        output_length
+            Number of time steps the torch module will predict into the future at once.
+        kernel_size
+            The size of every kernel in a convolutional layer.
+        num_filters
+            The number of filters in a convolutional layer of the TCN.
+        num_layers
+            The number of convolutional layers.
+        weight_norm
+            Boolean value indicating whether to use weight normalization.
+        dilation_base
+            The base of the exponent that will determine the dilation on every level.
+        dropout
+            The dropout rate for every convolutional layer.
+
+        Inputs
+        ------
+        x of shape `(batch_size, input_length, input_size)`
+            Tensor containing the features of the input sequence.
+
+        Outputs
+        -------
+        y of shape `(batch_size, input_length, 1)`
+            Tensor containing the predictions of the next 'output_length' points in the last
+            'output_length' entries of the tensor. The entries before contain the data points
+            leading up to the first prediction, all in chronological order.
+        """
+
+        super(_TCNModule, self).__init__()
+
+        # Defining parameters
+        self.input_size = input_size
+        self.input_length = input_length
+        self.n_filters = num_filters
+        self.kernel_size = kernel_size
+        self.out_len = output_length
+        self.dilation_base = dilation_base
+        self.dropout = nn.Dropout(p=dropout)
+
+        # If num_layers is not passed, compute number of layers needed for full history coverage
+        if (num_layers is None and dilation_base > 1):
+            num_layers = math.ceil(math.log((input_length - 1) / (kernel_size - 1), dilation_base))
+        elif (num_layers is None):
+            num_layers = (input_length - kernel_size) / 2 + 1
+
+        # Building TCN module
+        self.res_blocks_list = []
+        for i in range(num_layers):
+            res_block = _ResidualBlock(num_filters, kernel_size, dilation_base,
+                                       self.dropout, weight_norm, i, num_layers)
+            self.res_blocks_list.append(res_block)
+        self.res_blocks = nn.ModuleList(self.res_blocks_list)
```

Comment:
Is this needed?

Reply:
If I don't add this line the optimizer will not find the parameters, I get a 'optimizer got an empty parameter list' error.

Reply:
Oh! I think there should be a neater way to specify the parameters to the optimizer directly with pytorch

Reply:
Ok I'll look into it!